Implement difficulty tooltips for multiplayer lobbies

smallketchup82 opened the pull request:
Most of the discussion happened in #11306, and the issue can be closed if this is merged. This PR effectively adds difficulty information to the tooltip of maps in multiplayer lobbies.

Here is a preview of what it looks like:
![image](https://github.com/ppy/osu/assets/69545310/0fe1cabc-5106-4bbe-bd09-c32d5493b009)

The difficulty statistics update based on the mods applied, with the exception of the star rating (see below). I've done testing and found that it works regardless of whether the map is downloaded or not. Peppy said that this should be fine for the time being, until a new design is implemented. Which I can agree with.

### Known issues
- The star rating does not update based on the mods applied. I tried to implement this with `difficultyCache.getDifficultyAsync()` but to no avail. In the end I wasn't able to find a way to get the star rating to update, so I'll be leaving it for another contributor to implement. I suggest a new issue is opened for this if the PR gets merged.

smallketchup82 committed:
Implement difficulty statistics

smallketchup82 committed:
rollback my accidental global.json change

smallketchup82 committed:
Make the difficulty stats change based on the currently applied mods

smallketchup82 committed:
peppy said he prefers the version without pipes so ill remove the pipes

smallketchup82 committed:
peppy said he prefers the version without pipes so ill remove the pipes

bdach commented on osu.Game/Beatmaps/Drawables/DifficultyIcon.cs:
```diff
@@ -58,11 +61,14 @@ public virtual Bindable<StarDifficulty> Current
         /// Creates a new <see cref="DifficultyIcon"/>. Will use provided beatmap's <see cref="BeatmapInfo.StarRating"/> for initial value.
         /// </summary>
         /// <param name="beatmap">The beatmap to be displayed in the tooltip, and to be used for the initial star rating value.</param>
+        /// <param name="mods">The mods type beat</param>
```
please don't put memes in xmldoc thanks

smallketchup82 replied:
Completely missed this, sorry about that!

bdach commented on osu.Game/Beatmaps/Drawables/DifficultyIcon.cs:
```diff
@@ -58,11 +61,14 @@ public virtual Bindable<StarDifficulty> Current
+        /// <param name="mods">The mods type beat</param>
         /// <param name="ruleset">An optional ruleset to be used for the icon display, in place of the beatmap's ruleset.</param>
-        public DifficultyIcon(IBeatmapInfo beatmap, IRulesetInfo? ruleset = null)
+        public DifficultyIcon(IBeatmapInfo beatmap, IRulesetInfo? ruleset = null, Mod[]? mods = null)
```
Changing this class means changing every usage of it and not just multiplayer. Notably this icon & its tooltip are used on song select.

![osu_2024-01-17_13-44-26](https://github.com/ppy/osu/assets/20418176/9616c4ae-c2bf-47b9-8038-b94c5d461b2b)

I'm not super sure we want all of that stuff there. Certainly wasn't discussed.

smallketchup82 replied:
Also missed this, my bad. I think I'll just add another bool argument which defaults to false. It would be used to display all of this information, or just the star rating.

bdach commented on osu.Game/Beatmaps/Drawables/DifficultyIconTooltip.cs:
```diff
@@ -70,8 +160,43 @@ public void SetContent(DifficultyIconTooltipContent content)
+
+            // Misc row
+            length.Text = "Length: " + TimeSpan.FromMilliseconds(displayedContent.BeatmapInfo.Length / rate).ToString("mm\\:ss");
+            bpm.Text = " BPM: " + bpmAdjusted;
```
No rounding or significant digit spec, will look broken

smallketchup82 replied:
Beatmap length is effectively rounded via the formatting (the toString). This constrains it to be MM:SS.
As for the bpm, do you think it would be better to round to the 2nd decimal place, or just to whole numbers?

bdach replied:
Probably round to whole numbers to match song select.

smallketchup82 committed:
Apply suggested changes

- Change difficultyicon mods parameter docstring to be more professional
- Add a parameter for controlling whether the difficulty statistics show or not. Defaults to false
- Round the BPM in the tooltip to make sure it displays correctly

bdach commented on osu.Game/Beatmaps/Drawables/DifficultyIcon.cs:
```diff
@@ -58,11 +63,16 @@ public virtual Bindable<StarDifficulty> Current
         /// <param name="ruleset">An optional ruleset to be used for the icon display, in place of the beatmap's ruleset.</param>
-        public DifficultyIcon(IBeatmapInfo beatmap, IRulesetInfo? ruleset = null)
+        /// <param name="showTooltip">Whether to display a tooltip on hover. Defaults to false.</param>
+        public DifficultyIcon(IBeatmapInfo beatmap, IRulesetInfo? ruleset = null, Mod[]? mods = null, bool showTooltip = false)
```
`ShowTooltip` already existed on this class as a public property. Now there is `ShowTooltip` (the public property) and `showTooltip` (the constructor argument), and the former actually decides whether to show the tooltip, but the latter is passed to `DifficultyIconTooltipContent`, wherein it is stored to `ShowTooltip`, and used *not* to determine whether to *show* itself, but rather to show the *extended* content.

Please help. This just cannot be.

smallketchup82 replied:
I've refactored the name of the variable & constructor to be more descriptive and more accurate. I've also updated the docstring for `DifficultyIcon` to accurate reflect what the variable is used for.

So that we're on the same page, there isn't an issue with the implementation here right? My intent is to have the parameter in `DifficultyIcon`'s constructor determine whether the extended details are shown or not. The rest is just passing that to `DifficultyIconTooltipContent` to be put to use. Also, are you suggesting that the private property should have a different name than the constructor argument?

bdach replied:
> there isn't an issue with the implementation here right? My intent is to have the parameter in `DifficultyIcon`'s constructor determine whether the extended details are shown or not

I guess not. I'm in two minds as to whether it should be a bool flag or a subclass, however.



> Also, are you suggesting that the private property should have a different name than the constructor argument?

Definitely yes.

smallketchup82 replied:
> I'm in two minds as to whether it should be a bool flag or a subclass, however.

That's also a good way to go about it. I wouldn't mind implementing a bool or subclass. Whichever is cleaner is probably best.

> Definitely yes.

Got it

smallketchup82 replied:
Decided to just copy the same implementation as `ShowTooltips` and make `ShowExtendedTooltip` a public field. 

It doesn't really feel right when the way for determining if the tooltip is visible is by modifying the public field, while the way for determining if the content of the tooltip is extended is by passing a constructor argument.

smallketchup82 replied:
I'm happy enough with the current implementation, so I'll mark this as resolved

smallketchup82 committed:
Refactor the extended tooltip variables to be more descriptive

smallketchup82 committed:
Switch from using a constructor argument to using a public field for ShowExtendedTooltip

smallketchup82 commented on osu.Game/Beatmaps/Drawables/DifficultyIconTooltip.cs:
```diff
@@ -70,8 +153,50 @@ public void SetContent(DifficultyIconTooltipContent content)
             difficultyName.Text = displayedContent.BeatmapInfo.DifficultyName;
+
+            // Don't show difficulty stats if showExtendedTooltip is false
+            if (!displayedContent.ShowExtendedTooltip) return;
+
+            // Show the difficulty stats if showExtendedTooltip is true
+            difficultyFillFlowContainer.Show();
+            miscFillFlowContainer.Show();
```
The alpha on the elements are set to 0 by default. This seems a bit redundant when they're already hidden. The alpha value should be set to 1 only if showExtendedTooltip is true.

peppy replied:
Yes, but `SetContent` can be called *multiple times*. The whole premise behind these tooltips being structured as they are is that we can update the content via `SetContent` without recreating the full tooltip, allowing less re-initialisation but also animating between different content.

So imagine a sequence:

`SetContent(expanded:true)`
`SetContent(expanded:false)`

Please add test coverage to allow testing this and you'll see what I mean immediately.

smallketchup82 replied:
Ah I get it now. Will definitely add tests to be able to visualize this better.

smallketchup82 replied:
Applied the suggested changes. The tests show how the expanded content continues to be visible despite ShowExpandedTooltip being false. Thanks for pointing it out!

I've also added testing coverage for DifficultyIcon. Not sure If I placed it in the right directory, but I can easily move it somewhere else. If the location of the test isn't an issue, this can be marked as resolved.

smallketchup82 committed:
Add tests for DifficultyIcon

smallketchup82 committed:
Fix extended tooltip content still being shown despite ShowExtendedTooltip being false

smallketchup82 committed:
Conform to code style & remove unused imports

peppy committed:
Make test actually test multiple icons

peppy commented on osu.Game/Beatmaps/Drawables/DifficultyIconTooltip.cs:
```diff
@@ -70,8 +153,55 @@ public void SetContent(DifficultyIconTooltipContent content)
+            // Misc row
+            length.Text = "Length: " + TimeSpan.FromMilliseconds(displayedContent.BeatmapInfo.Length / rate).ToString("mm\\:ss");
+            bpm.Text = " BPM: " + Math.Round(bpmAdjusted, 0);
+            maxCombo.Text = " Max Combo: " + displayedContent.BeatmapInfo.TotalObjectCount;
```
This is not correct.

smallketchup82 replied:
Honestly looks fine without it. Guess I'll just remove it.
<img width="287" alt="Screenshot 2024-01-22 at 8 50 11 AM" src="https://github.com/ppy/osu/assets/69545310/a48a0503-d00a-4f1d-8b9f-d84ceedcb399">

bdach replied:
> I don't really feel like implementing it here

it's like two inches away

https://github.com/ppy/osu/blob/49c117793c7424fbbac10057f33f4b7d60a57fab/osu.Game/Beatmaps/StarDifficulty.cs#L19-L22

smallketchup82 replied:
> > I don't really feel like implementing it here
> 
> it's like two inches away
> 
> https://github.com/ppy/osu/blob/49c117793c7424fbbac10057f33f4b7d60a57fab/osu.Game/Beatmaps/StarDifficulty.cs#L19-L22

I remember trying to use that, but it didn't work so I used objectcount thinking it was the same.
<img width="295" alt="image" src="https://github.com/ppy/osu/assets/69545310/5c1c5a17-db65-43f5-8b9c-5b0576cbeec6">
`maxCombo.Text = " Max Combo: " + displayedContent.Difficulty.Value.MaxCombo;`

smallketchup82 committed:
Remove max combo reading & remove unnecessary commas

peppy committed:
Remove pointless comments

peppy committed:
Remove transparency

peppy committed:
Nullability

peppy committed:
Tidy things up

peppy committed:
Merge branch 'master' into multiplayer-difficulty-tooltip

peppy commented:
There's quite a bit of duplicated code here but we can deal with it later.

bdach committed:
Enable NRT in `TestSceneDifficultyIcon`

bdach committed:
Substitute two jank interdependent bool flags for single tri-state enums

bdach committed:
Fix test scene not properly setting tooltip type on all icons

bdach committed:
Merge branch 'master' into multiplayer-difficulty-tooltip

bdach committed:
Swap AR and OD on tooltips

Matches everything else.
